bittide-instances/src/Bittide/Instances/Hitl/SwCcTopologies.hs

```diff
@@ -0,0 +1,1057 @@
+      _ -> cur
+  testReset = unsafeFromActiveHigh testResetBool
+
+  -- Workaround for tests not resetting properly???
+  syncNodeProbablyWorking = changepoints sysClk testReset enableGen syncIn
+  syncNodeEnteredReset =
+    trueFor
+      (SNat @(Milliseconds 25))
+      sysClk
+      testReset
+      (not <$> syncNodeProbablyWorking)
+  syncNodePrevEnteredReset =
+    sticky sysClk testReset syncNodeEnteredReset
+
+  syncIn' = mux syncNodePrevEnteredReset syncIn (pure True :: Signal Basic125 Bool)
+  testCounter =
+    regEn
+      sysClk
+      sysRst
+      enableGen
+      (0 :: Unsigned 4)
+      (isFalling sysClk sysRst enableGen False startTest)
+      (satSucc SatBound <$> testCounter)
+
+  cfg = fromMaybe disabled <$> testConfig
+
+  ( txns
+    , txps
+    , swFincFdecs
+    , _callistoResult
+    , _callistoReset
+    , _dataCounts
+    , _stats
+    , spiDone
+    , spiOut
+    , transceiversFailedAfterUp
+    , allReady
+    , allStable
+    , clockShift
+    , allUgnsStable
+    , noFifoOverflows
+    , noFifoUnderflows
+    ) =
+      topologyTest
+        refClk
+        sysClk
+        ilaControl{skipTest = skip}
+        rxns
+        rxps
+        miso
+        cfg
+        calibratedClockShift
+
+  captureFlag =
+    riseEvery
+      sysClk
+      syncRst
+      enableGen
+      (SNat @(PeriodToCycles Basic125 (Milliseconds 1)))
+
+  milliseconds1 =
+    regEn
+      sysClk
+      syncRst
+      enableGen
+      (0 :: Unsigned 16)
+      captureFlag
+      (satSucc SatBound <$> milliseconds1)
+
+  tleDebugIla :: Signal Basic125 ()
+  tleDebugIla =
+    setName @"tleDebugIla"
+      ila
+      ( ilaConfig
+          $ "trigger_tle_0"
+          :> "capture_tle_0"
+          :> "probe_tle_milliseconds"
+          :> "probe_ilacfg_allReady"
+          :> "probe_ilacfg_startTest"
+          :> "probe_ilactl_syncStart"
+          :> "probe_startTest"
+          :> "probe_allStable"
+          :> "probe_tle_allUgnsStable"
+          :> "probe_tle_noFifoOverflows"
+          :> "probe_tle_noFifoUnderflows"
+          :> "probe_endSuccess"
+          :> "probe_startBeforeAllReady"
+          :> "probe_tle_transceiversFailedAfterUp"
+          :> "probe_testDone"
+          :> "probe_testSuccess"
+          :> Nil
+      )
+        { depth = D16384
+        }
+      sysClk
+      syncNodePrevEnteredReset
+      captureFlag
+      milliseconds1
+      allReady
+      startTest
+      syncStart
+      startTest
+      allStable
+      allUgnsStable
+      noFifoUnderflows
+      noFifoOverflows
+      endSuccess
+      startBeforeAllReady
+      transceiversFailedAfterUp
+      testDone
+      testSuccess
+
+  -- Check that tests are not synchronously started before all transceivers are up
+  startBeforeAllReady =
+    sticky
+      sysClk
+      syncRst
+      (syncStart .&&. ((not <$> allReady) .||. transceiversFailedAfterUp))
+
+  calibratedClockShift = capturedOnce
+   where
+    -- Possibly overkill.
```

Comment:
So actually I think it's worth bringing this up here since I wanted to ask if there was a good way to latch a value exactly once on a given signal. I think what I wrote there might be a bit over-engineered, and I'm also not sure if there's just a utility function somewhere that I overlooked.

Reply:
How about:
```hs
capturedOnce = 
  regEn clk rst ena Nothing 
  (isNothing <$> capturedOnce) 
  (orNothing <$> captureCond <*> capturedValue )
```
This samples `capturedValue` once, you can then use `fromMaybe defaultValue` to output `defaultValue` to set it to `defaultValue` untill you capture.

You then get
```hs
capturedOnce = fromMaybe defaultValue <$>
  regEn clk rst ena Nothing 
  (isNothing <$> capturedOnce) 
  (orNothing <$> captureCond <*> capturedValue )
  ```